### PR TITLE
Make propTypes optional for ComponentInterface

### PR DESCRIPTION
**Pull request by admmasters**
When creating native modules - required propTypes throws a false positive error.

Please fill in this template.

- [x] Use a meaningful title for the pull request. Include the name of the package modified.
- [x] Test the change in your own code. (Compile and run.)
- [x] Follow the advice from the [readme](https://github.com/DefinitelyTyped/DefinitelyTyped/blob/master/README.md#make-a-pull-request).
- [x] Avoid [common mistakes](https://github.com/DefinitelyTyped/DefinitelyTyped/blob/master/README.md#common-mistakes).
- [x] Run `npm run lint package-name` (or `tsc` if no `tslint.json` is present).

Select one of these and delete the others:

If changing an existing definition:
- [ ] Provide a URL to documentation or source code which provides context for the suggested changes: <<url here>>
- [ ] Increase the version number in the header if appropriate.
- [ ] If you are making substantial changes, consider adding a `tslint.json` containing `{ "extends": "dtslint/dt.json" }`.

### CODE REVIEW

**Review by iRoachie — COMMENTED** [review]
@admmasters It's a generic, can't you just pass in `any`? `ComponentInterface<any>` ?

**Comment by admmasters** [comment]
@iRoachie that doesn't work I'm afraid, but even if it did. Its incorrect in my opinion as propTypes are not required and are in fact not being used internally by Facebook (who are using Flow).

**Comment by iRoachie** [comment]
PropTypes are required definitions on the component class. If there’s no definitions for props then typescript can’t infer what this.props is.

export default class Test extends React.Component<any, any>

**Comment by admmasters** [comment]
```
export interface IProps extends ReactProps<any> {
  example: string;
}

export default class NativeView extends Component<IProps, any> {
  render() {
    return < NativeComponent {...this.props} />;
  }
}

const MyNativeView = requireNativeComponent('NativeView', NativeView);
```

**Comment by huhuanming** [comment]
Maybe you could do like this.

`
import { NativeModules } from 'react-native'

interface ICustomView {
   blabla...
}

const CustomView: ICustomView  = NativeModules.CustomView
`

**Review by alloy — CHANGES_REQUESTED** [review]
At the very least, even if it’s just for clear communication purposes, this PR could do with a test case that demonstrates the issue.

(I’m also marking this review as ‘request changes’ in order to silence the bot.)

**Comment by alloy** [comment]
I’m confused about the use-case of @admmasters and the conflation in the discussion between ‘PropTypes’ and ‘interface’ (i.e. TS types).

> its incorrect in my opinion as propTypes are not required and are in fact not being used internally by Facebook (who are using Flow).

PropTypes _for components that wrap native components_ are definitely required by the React Native runtime. This is a separate discussion from compile time type checking.

> I'm assuming this is required in the interface because: "we need to document it in propTypes (or we'll get an error that the native prop is undocumented" from React Native documentation?

Yes and also:

“This allows the infrastructure to verify that the propTypes match the native props to reduce the chances of mismatches between the ObjC and JS code.”

And:

“Since you don't want these native only properties to be part of the API, you don't want to put them in propTypes, but if you don't you'll get an error. The solution is simply to call them out via the nativeOnly option […]”

In short, the `propTypes` static property of a component class is **required** for components that wrap native components, so that the React Native runtime can verify that properties exported by the JS wrapper component match those exported from the native component. This is done in an effort to minimize hard to debug issues in native code.

Does this clear up the situation you’re running into @admmasters? If not, could you elaborate on your use-case, how it breaks, and why you think it should work that way?

**Comment by mhegazy** [comment]
Thanks for your contribution. This PR has unaddressed reviewer comments and can not be merged in at the time being. For housekeeping purposes we are closing stale PRs. If you'd still like to merge this code in, please open a new PR that has been merged and rebased with the `master` branch.
